Optimized various point-related :class:`.VMobject` methods

chopan050 opened the pull request:
<!-- Thank you for contributing to Manim! Learn more about the process in our contributing guidelines: https://docs.manim.community/en/latest/contributing.html -->

## Overview: What does this pull request change?
<!-- If there is more information than the PR title that should be added to our release changelog, add it in the following changelog section. This is optional, but recommended for larger pull requests. -->
<!--changelog-start-->
Main changes:
- Aggresively removed every call to `np.append` in `VMobject`'s methods, and reduced the use of `VMobject.append_points` as much as possible, because their extensive use is expensive. When possible, they're replaced with preallocations of empty ndarrays whose lengths can be precalculated.
- Modified `VMobject.append_points` to use `np.empty` instead of `np.append`, because it is slightly faster this way. Useful because this method can still get called a lot.
- Removed calls to `np.linspace` from `VMobject`'s methods, because repeated call to this method for small arrays is also too expensive.
- Added a new attribute `VMobject.bezier_alphas` related to the `nppcc`. This attribute is a precalculation of the alphas needed to interpolate pairs of anchors to obtain the required handles for the new straight segments that shall be represented by Bézier curves. This replaces those special cases where `np.linspace` was originally used for interpolation.
- Optimized `VMobject.consider_points_equals` by replacing the `np.allclose` method, which is also really expensive when used repeatedly to compare only two 3D points.
  - It is now implemented in a similar fashion to `VMobject.consider_points_equals_2d`, where a comparison coord-by-coord is made to determine if the points are close or not.
  - Both functions now consider two separate cases: when there's a single pair of points p0 and p1 (return a bool), and when there are multiple pairs of points (return a boolean ndarray).
    - Both implementations are much faster than the original `np.allclose` approach.
  - NOTE: I removed the "rtol". Let's discuss whether this is the best decision, please.
- <del>Added two new methods, `VMobject.consider_points_different` and `VMobject.consider_points_different_2d`, because more often than not we actually want to know if two points are distant enough from each other, rather than if they're close to each other.</del> _EDIT: Nevermind, I removed those methods to avoid code duplication._
- Added new methods `VMobject.get_subpath_split_indices` and `VMobject.get_subpath_split_indices_from_points`, which instead of explicitly obtaining a Python list of subpaths, it obtains an (n_subpaths, 2) ndarray of indices indicating where every subpath starts and ends.
  - These methods also accept parameters `n_dims` (which allows us to choose between 2D or 3D point comparison) and `strip_null_end_curves` (necessary for `VMobject.align_points` where the null end curves at every subpath must be removed as a fix to certain bug).
- Rewrote `VMobject.get_subpaths_from_points` and `VMobject.gen_subpaths_from_points_2d` to use this new method `VMobject.get_subpath_split_indices_from_points`.
  - `VMobject._gen_subpaths_from_points` is no longer used and can probably be deprecated if this PR is accepted.
- Some methods were completely rewritten:
  - `VMobject.add_points_as_corners`: it originally used a for loop to add every point one by one through `VMobject.add_line_to`. This was extremely expensive (especially with the previous use of `np.linspace`) and was completely changed to just add all the points at once, calculating the necessary handles via interpolations.
  - `VMobject.change_anchor_mode`: if using "jagged" mode, it is not necessary at all to separate the points into subpaths (a simple interpolation is sufficient). Only get the subpaths in "smooth" mode. In this case, the new implementation now uses this new method `VMobject.get_subpath_split_indices` instead of explicitly precalculating all the subpaths previously and storing them all in a Python list. As the number of points remains pretty much the same and only the handles change, there's no need to clear the points or create another ndarray of points: we can just directly overwrite the handles.
  - `VMobject.align_points`: this method was completely rewritten so that it now uses `VMobject.get_subpath_indices` too. In this way, one can also easily calculate useful information like how long is every subpath (just subtract the pairs of indices!) and, thus, which subpath is longer when comparing pairs of subpaths between the two `VMobject`s (via `np.maximum`). The `np.append`s were obliterated. As mentioned earlier, this new implementation uses `strip_null_end_curves=True` to remove the excess null curves at the end of the subpaths, if any.
- Introduced the concept of `Mobject.memory`, which might be useful for memoizing certain desirable properties.
  - Used it at `VMobject.point_from_proportion`.
    - My main motivation was that I first managed to optimize (via binary search through the accumulated Bézier lengths) the process of finding the Bézier curve from where to calculate the point from the given proportion. This would mean going from an O(n) algorithm to an O(log(n)) algorithm... except that it still needed to calculate all the Bézier curve lengths in O(n) operations. If only we didn't need to calculate them...
    - Created two new methods, `VMobject._init_curve_memory` and `VMobject._update_curve_memory`, which precalculate the lengths and accumulated lengths of the Bézier curves making up the `VMobject` and update it only if the actual points or the number of sample points per curve have changed.
    - This all resulted in a huge improvement in `VMobject.point_from_proportion`'s time!
    - Didn't go much further with this idea because I first wanted your opinions about this. There are more potential applications for this idea. For example, there's a method in ManimGL, `VMobject.get_area_vector`, where this could be used as well.
- Some other (minor?) changes.

<!--changelog-end-->

## Motivation and Explanation: Why and how do your changes improve the library?
<!-- Optional for bugfixes, small enhancements, and documentation-related PRs. Otherwise, please give a short reasoning for your changes. -->
My original intention was to optimize my test scene where I have 110 ParametricFunctions being updated in a span of 5 seconds, which is really expensive!

```py
from manim import *

class LaplaceScene(Scene):
    def construct(self):
        axes = Axes(x_range=[-1, 7], y_range=[-1, 4], x_length=16.32, y_length=10.20).add_coordinates()
        s = ValueTracker(0)
    
        def get_min_x(k):
            if k > 3:
                return np.floor(2*np.log(k/3.5)/0.5) / 2 - 0.5
            if k < 0:
                return np.floor(2*np.log(-k/0.5)/0.5) / 2 - 0.5
            return -0.5

        draw_horizontal = lambda k: axes.plot(
            lambda t: k*np.exp(-s.get_value()*t), x_range=[get_min_x(k), 6.5, 0.5],
            stroke_width=1.0, stroke_color=RED_A,
            use_vectorized=True,
        )
        horizontals = {k: draw_horizontal(k) for k in range(-15, 96)}
        del horizontals[0]

        verticals = {
            k: Line(
                axes.c2p(k, -1), axes.c2p(k, 5),
                stroke_width=1.0, stroke_color=BLUE_A,
            )
            for k in range(1, 7)
        }

        draw_func = lambda: axes.plot(
            lambda t: t*np.exp(-s.get_value()*t), x_range=[0, 6.5, 0.5],
            stroke_color=YELLOW,
            use_vectorized=True,
        )
        func = draw_func()

        self.play(
            *[Create(verticals[key]) for key in verticals],
            *[Create(horizontals[key]) for key in horizontals],
            DrawBorderThenFill(axes),
        )
        self.play(Create(func), run_time=2)

        func.add_updater(lambda plot: plot.become(draw_func()))
        for k in horizontals:
            (lambda k: horizontals[k].add_updater(
                lambda line: line.become(draw_horizontal(k))
            ))(k)

        self.play(s.animate.set_value(0.5), run_time=5)

        func.clear_updaters()
        for k in horizontals:
            horizontals[k].clear_updaters()

        self.wait()
        self.play(VGroup(axes, func, *horizontals.values(), *verticals.values()).animate.scale(0.5))
        self.wait()
        
with tempconfig({"preview": True, "quality": "medium_quality", "disable_caching": True}):
    scene = LaplaceScene()
    scene.render()
```

Almost half of the time of the scene is spent on `ParametricFunction.generate_points`, and there are 3 major culprits: `VMobject.make_smooth`, `VMobject.add_points_as_corners`, and `Axes.coords_to_point`.

![imagen](https://github.com/ManimCommunity/manim/assets/49853152/ee2b3c8c-03e0-471e-b897-726bd8e2bc42)

In #3281 I optimized `get_smooth_handle_points` (and other Bézier-related functions), which was an important cause of `VMobject.make_smooth`'s excessive time, and in #3285 and #3286 I addressed the issue with `Axes.coords_to_point` and, more specifically, `NumberLine.number_to_point`.

Regarding the `VMobject.add_points_as_corners` and the `VMobject.get_subpaths`, which are the other major issues, I initially rewrote `ParametricFunction` to circumvent those issues. After all, I don't think `ParametricFunction.generate_points` should be relying a lot on these functions to, well, generate its points, when they can be generated in other efficient ways. I have some ideas about rewriting `ParametricFunction`, which I wanna discuss in a future PR.

However, there are still many other `VMobject` subclasses which depend on those currently inefficient methods too. As I noticed this was a general issue for many `VMobject`s, I decided to address the root issues and directly modify and optimize most of `VMobject`'s point-related methods.

**EDIT: here's a new screenshot of how `ParametricFunction.generate_points` looks after these changes. `VMobject.get_subpaths` (replaced by `VMobject.get_subpath_split_indices`) and `VMobject.add_points_as_corners` are no longer a problem!**
![imagen](https://github.com/ManimCommunity/manim/assets/49853152/db9d3234-283d-41b2-a3b6-7c642c469263)

## Links to added or changed documentation pages
<!-- Please add links to the affected documentation pages (edit the description after opening the PR). The link to the documentation for your PR is https://manimce--####.org.readthedocs.build/en/####/, where #### represents the PR number. -->


## Further Information and Comments
<!-- If applicable, put further comments for the reviewers here. -->



<!-- Thank you again for contributing! Do not modify the lines below, they are for reviewers. -->
## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

behackl commented on manim/mobject/types/vectorized_mobject.py:
```diff
@@ -1096,6 +1260,134 @@ def get_subpaths(self) -> tuple:
+        n_curves = len(ends)
+        starts = starts[:n_curves]
+
+        # DO NOT DELETE THIS! Or else, a [[0 0]] ndarray will be returned
+        # for an empty list of points, when it should instead return [].
```
The latter part of the comment suggests that an empty Python list is returned, which is not the case. The wording should be improved.

chopan050 replied:
I see. I'm going to check this as well.

chopan050 replied:
Done!

behackl commented on manim/mobject/types/vectorized_mobject.py:
```diff
@@ -1096,6 +1260,134 @@ def get_subpaths(self) -> tuple:
+            is_equal = self.consider_points_equals
+            is_different = self.consider_points_different
+
+        is_far = is_different(starts[1:], ends[:-1])
```
Semantics: I would have expected `is_far` to be a function (just like `is_different`, `is_equal`), rather than an array -- it might be worth renaming the variable.

chopan050 replied:
Good point. That also bothered me a bit, but I didn't know which names could be better. I'll take a deeper look at this issue.

chopan050 replied:
Fixed! The new names should no longer cause confusion IMO.

behackl commented on manim/mobject/types/vectorized_mobject.py:
```diff
@@ -1096,6 +1260,134 @@ def get_subpaths(self) -> tuple:
+        )
+
+        self.memory["piece_curves"] = {
+            "points": self.points,
```
Do you really need to store another reference to `self.points` inside of `self.memory`?

chopan050 replied:
My idea was to store the previous state of `self.points`. In this way, the Bézier curve lengths would be recalculated only if the points change.

Now... "reference"? Oh my, I haven't noticed that! This should be a copy of `self.points`, not a reference! Thanks for bringing this up.

chopan050 replied:
Done! I addressed this issue: now `self.memory["piece_curves"]["points"]` actually holds the previous state of `self.points`, instead of an useless reference.
I also fixed some other details I missed earlier regarding `VMobject._update_curve_memory`.

behackl commented on manim/mobject/types/vectorized_mobject.py:
```diff
@@ -1282,24 +1572,38 @@ def point_from_proportion(self, alpha: float) -> np.ndarray:
+        num_curves = self.get_num_curves()
 
-                return curve(residue)
+        # Binary search
```
No need to implement this manually: use `bisect` from Python's default [bisect](https://docs.python.org/3.8/library/bisect.html) module.

chopan050 replied:
Oh nice, the more you know...
I'll check this out!

chopan050 replied:
I used `bisect.bisect_left` and it worked as a charm, it was even faster than my manual implementation!

behackl commented on manim/mobject/types/vectorized_mobject.py:
```diff
@@ -1402,13 +1706,28 @@ def get_anchors(self) -> np.ndarray:
-        )
+        num_curves = self.get_num_curves()
+        anchors = np.empty((2 * num_curves, self.dim))
+        anchors[0::2] = self.get_start_anchors()[:num_curves]
+        anchors[1::2] = self.get_end_anchors()
```
Why is the slice required for the start anchors but not the end anchors?

chopan050 replied:
Because I've seen in the source code that there are some cases where `self.points` ends with a single start anchor (or straight up has a single point in general) instead of a full quad representing a Bézier curve. In these cases, these lone points do not increase the `num_curves` and can be stripped off.

However, when there's an end anchor ending a Bézier curve, this also means there's another Bézier curve to consider in `num_curves`.

What I mean is, there are as many Bézier curves as there are end anchors, and thus there's no need to slice these end anchors. However, not every start anchor means the presence of another Bézier curve, and thus the start anchors **must** be sliced, just in case.

But yeah, this might be confusing for other people looking at the source code, so I should add an explanationy comment in the code (or straight up slice both arrays of anchors, even if the end anchors do not really need it).

chopan050 replied:
Added the corresponding comment.

behackl commented on manim/mobject/types/vectorized_mobject.py:
```diff
@@ -1646,41 +2021,61 @@ def pointwise_become_partial(
+        # The following two lines will compute which Bézier curves of the given Mobject must be processed.
+        # The residue indicates the proportion of the selected Bézier curve which must be selected.
+        #
+        # Example: if num_curves is 10, a is 0.34 and b is 0.78, then:
+        # - lower_index is 3 and lower_residue is 0.4, which means the algorithm will look at the 3rd Bezier
+        #   and select its part which ranges from t=0.4 to t=1.
+        # - upper_index is 7 and upper_residue is 0.8, which means the algorithm will look at the 7th Bezier
+        #   and select its part which ranges from t=0 to t=0.8.
```
Might be a good doc- or even unit test?

chopan050 replied:
That would be interesting.

behackl reviewed (changes requested):
Fantastic work, looks like a lot of well thought through optimizations. I have a bunch of questions and suggestions, please take a look -- given that the tests are more or less happy though, I'm sure we'll be able to get this merged rather sooner than later.

chopan050 commented on manim/mobject/types/vectorized_mobject.py:
```diff
@@ -883,26 +921,27 @@ def change_anchor_mode(self, mode: str):
+                starts = submob.get_start_anchors()
+                ends = submob.get_end_anchors()
+                for a in self.bezier_alphas:
+                    submob.points[1::nppcc] = interpolate(starts, ends, a)
```
Yes, my bad - going to correct that!

chopan050 replied:
Fixed! There was also some bad slicing going on, which I didn't notice before: `submob.points[1::nppcc]`, the first group of handles, was getting incorrectly reassigned every time, instead of assigning the handle groups at the correct positions! It should have been something like `submob.points[i::nppcc]`, with an `i` instead of an `1`. I ended up using a `for i in range(1, nppcc - 1):` to handle this.

chopan050 commented on manim/mobject/types/vectorized_mobject.py:
```diff
@@ -977,10 +1014,49 @@ def scale_handle_to_anchor_distances(self, factor: float):
+
+        # Case 2: multiple pairs of points
+        is_close = abs(p1 - p0) <= atol
+        return is_close[:, 0] & is_close[:, 1] & is_close[:, 2]
```
I've found that using methods like `np.all` or `np.sum` are actually slower when we only have to reduce 3 coordinates, so that's why I implemented this manually instead of doing something like `np.all(is_close, axis=1)` or `is_close.all(axis=1)`.

chopan050 replied:
Maybe I could add a comment in the code explaining this, though.

chopan050 replied:
Added some brief comments in the code regarding this.

behackl replied:
When you say "slower", how much slower are we actually talking?

chopan050 replied:
Well, using the same test scene:
- By using manual ANDs, `VMobject.consider_points_equals` takes 0.56% of the time, `VMobject.consider_points_equals_2d` has 1.13%, and in general `VMobject.get_subpath_split_indices_from_ponits` has 3.23%. 
![imagen](https://github.com/ManimCommunity/manim/assets/49853152/db5c6a68-09de-4791-92cb-5f57819bb7c6)

- By using `np.all`, `VMobject.consider_points_equals` takes 1.11% of the time, `VMobject.consider_points_equals_2d` has 1.76%, and in general `VMobject.get_subpath_split_indices_from_ponits` has 4.47%. 
![imagen](https://github.com/ManimCommunity/manim/assets/49853152/b2deaea3-fa64-40ca-8550-f0acd59602de)

And in general, by using `timeit` in a Jupyter Notebook I found out that using `np.all` takes twice as much time as using manual ANDs.

So, `VMobject.consider_points_equals` is 2x faster, `VMobject.consider_points_equals_2d` is around 1.5x faster, and `VMobject.get_subpath_split_indices_from_points` is around 1.25x faster.

So these three functions are considerably faster by themselves, although in the grand scale of things it doesn't influence too much in the render time of the whole scene.

chopan050 commented on manim/mobject/types/vectorized_mobject.py:
```diff
@@ -994,22 +1070,111 @@ def consider_points_equals_2d(self, p0: np.ndarray, p1: np.ndarray) -> bool:
+        p0 = p0.reshape(-1, self.dim)[:, :2]
+        p1 = p1.reshape(-1, self.dim)[:, :2]
+        is_close = abs(p1 - p0) <= atol
+        return is_close[:, 0] & is_close[:, 1]
```
Same as above: `np.all` is actually slower.

chopan050 commented on manim/mobject/types/vectorized_mobject.py:
```diff
@@ -994,22 +1070,111 @@ def consider_points_equals_2d(self, p0: np.ndarray, p1: np.ndarray) -> bool:
+        is_close = abs(p1 - p0) <= atol
+        return is_close[:, 0] & is_close[:, 1]
+
+    def consider_points_different(
```
Hmm, I mainly did this to:
- avoid more function calls which add a little bit of overhead
- avoid negating the result of `consider_points_equals`, because it is also slightly faster to directly compute if the two points are different, rather than computing if they're equal, then negating the result

Both methods could be used everywhere: `consider_points_different` is crucial for splitting points into subpaths, and `consider_points_equals` is necessary for methods like `Camera.set_cairo_context_path` which checks if the subpaths should be closed.

As these methods are essential, one would want them to be as fast as possible, and that's why I did that. However, I'm myself also bothered by the duplication of code, and I'm wondering if the small gains in speed are **really** worth it.

chopan050 replied:
In the end, I decided to get rid of `VMobject.consider_points_different` and `VMobject.consider_points_different_2d`, and just use the negation of their "equals" counterparts.

chopan050 commented on manim/mobject/types/vectorized_mobject.py:
```diff
@@ -994,22 +1070,111 @@ def consider_points_equals_2d(self, p0: np.ndarray, p1: np.ndarray) -> bool:
+        is_far = abs(p1 - p0) > atol
+        return is_far[:, 0] | is_far[:, 1] | is_far[:, 2]
+
+    def consider_points_different_2d(
```
I left a reply above regarding this situation.

chopan050 commented on manim/mobject/types/vectorized_mobject.py:
```diff
@@ -1027,15 +1192,17 @@ def gen_cubic_bezier_tuples_from_points(self, points: np.ndarray) -> tuple:
     def get_cubic_bezier_tuples(self):
         return self.get_cubic_bezier_tuples_from_points(self.points)
 
+    # TODO: Deprecate?
```
Yeah, another PR would be fine for this.

chopan050 commented on manim/mobject/types/vectorized_mobject.py:
```diff
@@ -1096,6 +1260,134 @@ def get_subpaths(self) -> tuple:
+        starts = points[::nppcc]
+        ends = points[nppcc - 1 :: nppcc]
+        # This ensures that there are no more starts than ends.
+        # TODO: ends.shape[0] would be more efficient, but some test cases regarding
```
I'd like to collaborate if possible!
I have many doubts and questions regarding the format of these variables, like:
- Should the `points` parameters **always have to be ndarrays**?
- Should we enforce the use of `np.asarray` or `np.array` at the beginning of most methods to ensure that we're working with NumPy arrays rather than lists, or can we sometimes just assume that without using `np.asarray`?
- What methods should be 'more exposed' to the animators creating new scenes, and thus should be prepared in case the users decide to pass a Python list instead of a NumPy array?
- etc.

No need to reply here, I think this could be better discussed in another place.

chopan050 commented:
Done! I addressed most comments.
One thing I forgot to mention is regarding `VMobject.insert_n_curves_to_point_list`:

```py
    def insert_n_curves_to_point_list(self, n: int, points: np.ndarray) -> np.ndarray:
        # ... other stuff
        for quad, sf in zip(bezier_quads, split_factors):
            if sf == 1:
                new_points[start_i : start_i + nppcc] = quad
                start_i += nppcc
            else:
                for i in range(sf):
                    new_points[start_i : start_i + nppcc] = partial_bezier_points(
                        quad, i / sf, (i + 1) / sf
                    )
                    start_i += nppcc
        return new_points
```

In #3281 I added a new function in `beziers.py` called `subdivide_bezier`, and mentioned that if that PR gets approved it would be a good idea to replace the use of `partial_bezier_points` in the above code with `subdivide_beziers`:

```py
    def insert_n_curves_to_point_list(self, n: int, points: np.ndarray) -> np.ndarray:
        # ... other stuff
        for quad, sf in zip(bezier_quads, split_factors):
            new_points[start_i : start_i + sf * nppcc] = subdivide_bezier(quad, sf)
            start_i += sf * nppcc
        return new_points
```

This would look much cleaner and is also more efficient, since in `partial_bezier_points` we're splitting the Béziers twice to get the desired portion in the middle (discarding the start and end portions, if any), so using that function in `VMobject.insert_n_curves_to_point_list` is double the necessary work. On the other hand, with `subdivide_bezier` the work is done only once and all the pieces are used, no portion is discarded.

So, if #3281 gets merged first, I'd wanna add this change to this PR before merging this one.

chopan050 commented:
Closed this PR, because it was superseded by smaller PRs such as #3759, #3760 and #3765.
